In [2]:
from utils import styled_print, create_dir, extract_images, extract_paragraphs, \
    random_select_dict


In [1]:
%load_ext autoreload
%autoreload 2

In [3]:
book_file_path = "./data/fire-and-blood.docx"

## Extract Raw Data From the Docx File

Using Microsoft Word's Word Counter Tool, we found the following information:

| Property      | Count |
| ----------- | ----------- |
| Pages       | 703         |
| Words       | 256,032     |
| Paragraphs  | 3197        |
| Lines       | 23,849      |
| Graphics    | 71          |
| Characters (No Space) | 1,189,352 |
| Characters (With Space) | 1,441,453 |

We have used Microsoft Word's Advance Search Feature (Use flag `^g`) to find total number of graphical objects in our document.

The goal of this part of code is to extract similar amount of data from our document. 

**Note** Our numbers might not exactly match these statics but we will try to capture numbers as close to these stats. 

### Extract Images

In [4]:
image_dir = create_dir("./data", "images")
image_file_paths = extract_images(book_file_path, image_dir, verbose=0)
styled_print(f"Found Total {len(image_file_paths)} Images from the Book {book_file_path}", header=True)

› creating directory ... ./data/images
› Extracting Images from ./data/fire-and-blood.docx
› Found Total 112 Images from the Book ./data/fire-and-blood.docx


**Observations**
- As indicated here, we were able to find total 112 images from the book. 
- We have manually checked all the images to verify that there is no `Non-Image` object present. 
- We have found that our approach to extract images actually `extracted all the correct images` and it is much better that using Microsoft Word's Advance Search option to find images. 

### Extract Paragraphs

In [5]:
paragraphs = extract_paragraphs(book_file_path, min_char_count=1)
styled_print(f"Found Total {len(paragraphs)} Paragraphs from the Book {book_file_path}", header=True)

› Extracting Paragraphs from ./data/fire-and-blood.docx
› Found Total 3168 Paragraphs from the Book ./data/fire-and-blood.docx


In [6]:
styled_print(f"Some Sample Paragraphs from the Book {book_file_path}", header=True)
sampled_paragraphs = random_select_dict(paragraphs, 5)
for key, val in sampled_paragraphs.items():
    styled_print(f"{key} - {val}")


› Some Sample Paragraphs from the Book ./data/fire-and-blood.docx
    494 - It may be that conflict was inevitable, for the two brothers had very different natures. King Aenys loved his wife, his children, and his people, and wished only to be loved in turn. Sword and lance had lost whatever appeal they ever had for him. Instead His Grace dabbled in alchemy, astronomy, and astrology, delighted in music and dance, wore the finest silks, samites, and velvets, and enjoyed the company of maesters, septons, and wits. His brother, Maegor, taller, broader, and fearsomely strong, had no patience for any of that, but lived for war, tourneys, and battle. He was rightly regarded as one of the finest knights in Westeros, though his savagery in the field and his harshness toward defeated foes was oft remarked upon as well. King Aenys sought always to please; when faced with difficulties, he would answer
    1542 - Elsewhere in the Red Keep, Queen Alysanne spent the morning with her children; Prince